DESIGN.md

```diff
@@ -0,0 +1,400 @@
+    func (m *SignMessage) SigStructure(external []byte, signature *Signature) (ToBeSigned []byte, err error)
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) (err error)
+    func (message *SignMessage) UnmarshalCBOR(data []byte) (err error)
+    func (m *SignMessage) Verify(external []byte, verifiers []Verifier) (err error)
+type Signature
+    func NewSignature() (s *Signature)
+    func (s *Signature) Decode(o interface{})
+    func (s *Signature) Equal(other *Signature) bool
+type Signer
+    func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error)
+    func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Signer, err error)
+    func (s Signer) GetAlg() *Algorithm
+    func (s *Signer) Public() (publicKey crypto.PublicKey)
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm
+    AlgorithmPS384   *Algorithm
+    AlgorithmPS512   *Algorithm
+    AlgorithmES256   *Algorithm
+    AlgorithmES384   *Algorithm
+    AlgorithmES512   *Algorithm
+    AlgorithmEd25519 *Algorithm
+)
+
+type Algorithm struct {
+    Name     string
+    Value    int
+    HashFunc crypto.Hash
+}
+
+func I2OSP(x *big.Int, n int) ([]byte, error)
+func OS2IP(x []byte) *big.Int
+
+type ProtectedHeader map[interface{}]interface{}
+    func (h *ProtectedHeader) MarshalCBOR() ([]byte, error)
+    func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error
+    func (h *ProtectedHeader) SetAlgorithm(alg *Algorithm)
+
+type UnprotectedHeader map[interface{}]interface{}
+
+type Headers struct {
+    RawProtected   []byte
+    RawUnprotected []byte
+    Protected      ProtectedHeader
+    Unprotected    UnprotectedHeader
+}
+
+type Signature struct {
+    Headers   Headers
+    Signature []byte
+}
+    func NewSignature() *Signature
+    func (s *Signature) MarshalCBOR() ([]byte, error)
+    func (s *Signature) UnmarshalCBOR(data []byte) error
+
+type SignMessage struct {
+    Headers    Headers
+    Payload    []byte
+    Signatures []*Signature
+}
+    func NewSignMessage() *SignMessage
+    func ParseSignMessage(data []byte) (*SignMessage, error)

```

Comment:
AFAIU `ParseSignMessage` seems redundant. Isn't it the same as:

```go
  sm := new(SignMessage)
  sm.UnmarshalCBOR(data)
```

Reply:
Yes, `ParseSignMessage` is just for UX.

Basically,

```golang
func ParseSignMessage(data []byte) (*SignMessage, error) {
    var msg SignMessage
    if err := msg.UnmarshalCBOR(data); err != nil {
        return nil, err
    }
    return &msg, nil
}
```

Reply:
Hmm, I don't see the UX improvements, both approaches takes 3 lines, one is slightly less verbose: 

```go
var msg cose.SignMessage
if err := msg.UnmarshalCBOR(data); err != nil {
    return nil, err
}

// vs

msg, err := cose.ParseSignMessage(data)
if err != nil {
    return nil, err
}
```

But now there are two ways to do the same, which is confusing, and more APIs to maintain.

Feel free to ignore this opinion, as it might be a matter of taste.